src/core/Akka.Streams/Implementation/AsyncEnumerable.cs

```diff
@@ -0,0 +1,77 @@
+﻿// //-----------------------------------------------------------------------
+// // <copyright file="AsyncEnumerable.cs" company="Akka.NET Project">
+// //     Copyright (C) 2009-2021 Lightbend Inc. <http://www.lightbend.com>
+// //     Copyright (C) 2013-2021 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// // </copyright>
+// //-----------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Akka.Configuration.Hocon;
+
+namespace Akka.Streams.Dsl
+{
+    /// <summary>
+    /// Used to treat an <see cref="IRunnableGraph{TMat}"/> of <see cref="ISinkQueue{T}"/>
+    /// as an <see cref="IAsyncEnumerable{T}"/>
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public sealed class StreamsAsyncEnumerableRerunnable<T> : IAsyncEnumerable<T>
+    {
+        private readonly IMaterializer _materializer;
+        private readonly IRunnableGraph<(UniqueKillSwitch, ISinkQueue<T>)> _graph;
+
+        public StreamsAsyncEnumerableRerunnable(IRunnableGraph<(UniqueKillSwitch, ISinkQueue<T>)> graph, IMaterializer materializer)
+        {
+            _graph = graph;
+            _materializer = materializer;
+        }
+
+        public IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            return new SinkQueueAsyncEnumerator<T>(_graph.Run(_materializer), cancellationToken);
+        }
+    }
+    /// <summary>
+    /// Wraps a Sink Queue and Killswitch around <see cref="IAsyncEnumerator{T}"/>
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public sealed class SinkQueueAsyncEnumerator<T> : IAsyncEnumerator<T>
+    {
+        private ISinkQueue<T> _sinkQueue;
+        private IKillSwitch _killSwitch;
+        private CancellationToken _token;
+        public SinkQueueAsyncEnumerator((UniqueKillSwitch killSwitch,ISinkQueue<T> sinkQueue) queueAndSwitch, CancellationToken token)
+        {
+            _sinkQueue = queueAndSwitch.sinkQueue;
+            _killSwitch = queueAndSwitch.killSwitch;
+            _token = token;
```

Comment:
So, I'm honestly not sure if this is the right way to use the `CancellationToken` here or not.

Reply:
Added tests to confirm this is the right way.